torch/utils/module_tracker.py

```diff
@@ -74,6 +93,54 @@ def is_bw(self):
         """
         return torch._C._current_graph_task_id() != -1
 
+    def get_known_fqn(self, mod):
+        """
+        Return the fqn for the given module if it is known to the Module Tracker, otherwise None
+        """
+
+        if mod not in self._known_modules:
+            return None
+        return self._known_modules[mod]
+
+    def register_user_hooks(
+        self,
+        pre_fw_hook: Optional[Callable] = None,
+        post_fw_hook: Optional[Callable] = None,
+        pre_bw_hook: Optional[Callable] = None,
+        post_bw_hook: Optional[Callable] = None,
+    ):
+        """
+        Registers user specified hooks that are called before/after the forward/backward pass for
+        each module tracked by the Module Tracker. One or more can be none.
+
+        The hooks should have the following signatures.
+
+        pre_fw_hook (fqn, module, input) : Called before the forward pass for the module.
+        post_fw_hook (fqn, module, input, output) : Called after the forward pass for the module.
+        pre_bw_hook (fqn, *outputs) : A multi-grad hook that is called on all the outputs of the
+                                      module that require gradients.
+        post_bw_hook (fqn, *inputs) : A multi-grad hook that is called on all the inputs of the
+                                      module that require gradients.
+
+        They have the same signature, constraints and requirements as the hooks registered with
+        ``register_module_forward_pre_hook``, ``register_module_forward_hook`` and
```

Comment:
I am not sure about this. Some will allow modifying arguments and some won't.
If the user wants to change module inputs, they can register their own module_ hook.

Note: didn't comment on the impl below since I guess it's going to change.

Reply:
Each of these hooks have their own contract and we respect that in the module tracker.
pre-foward: input modification is allowed and will be passed to the module
post-forward: output modification is allowed and will be applied. Input modification is a nop
multi-grad: No modifications allowed

Reply:
My question is why would we do that as opposed to:
- These hooks are markers only. You should never return anything fro them.

Reply:
Okay so no modifications are allowed is the contract.